src/pipecat/processors/canonical_metrics_processor.py

```diff
@@ -0,0 +1,202 @@
+import os
+import uuid
+import wave
+from datetime import datetime
+from io import BytesIO
+from typing import Dict, List, Tuple
+
+import aiohttp
+from loguru import logger
+
+try:
+    import aiofiles
+    import aiofiles.os
+except ModuleNotFoundError as e:
+    logger.error(f"Exception: {e}")
+    logger.error(
+        "In order to use Canonical Metrics, you need to `pip install pipecat-ai[canonical]`. " +
+        "Also, set the `CANONICAL_API_KEY` environment variable.")
+    raise Exception(f"Missing module: {e}")
+
+
+from pipecat.frames.frames import CancelFrame, EndFrame, Frame
+from pipecat.processors.audio_buffer_processor import AudioBufferProcessor
+from pipecat.processors.frame_processor import FrameDirection
+
+"""
+This class extends AudioBufferProcessor to handle audio processing and uploading
+for the Canonical Voice API.
+"""
+
+
+class CanonicalMetrics(AudioBufferProcessor):
+    """
+    Initialize a CanonicalAudioProcessor instance.
+
+    This class extends AudioBufferProcessor to handle audio processing and uploading
+    for the Canonical Voice API.
+
+    Args:
+        call_id (str): Your unique identifier for the call. This is used to match the call in the Canonical Voice system to the call in your system.
+        assistant (str): Identifier for the AI assistant. This can be whatever you want, it's intended for you convenience so you can distinguish
+        between different assistants and a grouping mechanism for calls.
+        assistant_speaks_first (bool, optional): Indicates if the assistant speaks first in the conversation. Defaults to True.
+        output_dir (str, optional): Directory to save temporary audio files. Defaults to "recordings".
+        default_part_size (int, optional): Default size for multipart upload parts in bytes. Defaults to 1MB (1024 * 1024 * 1).
+
+    Attributes:
+        call_id (str): Stores the unique call identifier.
+        assistant (str): Stores the assistant identifier.
+        assistant_speaks_first (bool): Indicates whether the assistant speaks first.
+        output_dir (str): Directory path for saving temporary audio files.
+        partsize (int): Size of each part for multipart uploads.
+
+    The constructor also ensures that the output directory exists.
+    This class requires a Canonical API key to be set in the CANONICAL_API_KEY environment variable.
+    """
+
+    def __init__(
+            self,
+            call_id: str,
+            assistant: str,
+            assistant_speaks_first: bool = True,
+            output_dir: str = "recordings",
+            default_part_size: int = 1024 * 1024 * 1):
+        super().__init__()
+        if not os.environ.get("CANONICAL_API_KEY"):
+            raise ValueError(
+                "CANONICAL_API_KEY is not set, a Canonical API key is required to use this class")
```

Comment:
Could we pass `api_key: str` to the constructor instead? This is the approach we take in many other services. Form outside you can do `os.environ.get("CANONICAL_API_KEY")` or get it from somewhere else.

Reply:
Done.